core/src/Dscp/Crypto/MerkleTree.hs

```diff
@@ -85,80 +84,106 @@ instance Foldable MerkleTree where
+-- is hashed.
+mkMerkleSig :: HasHash a => a -> MerkleSignature a
+mkMerkleSig a = MerkleSignature 1 (unsafeHash a)
+
+-- | Makes a leaf node of Merkle tree given the element of type `a`.
+mkLeaf :: HasHash a => a -> MerkleNode a
+mkLeaf a = MerkleLeaf
+    { mnVal   = a
+    , mnRoot  = mkMerkleSig a
     }
 
+-- | Given two Merkle subtrees, makes a tree which has them
+-- as children.
 mkBranch :: MerkleNode a -> MerkleNode a -> MerkleNode a
 mkBranch l r = MerkleBranch
-    { mLeft  = l
-    , mRight = r
-    , mRoot  = mkBranchRootHash (mRoot l) (mRoot r)
+    { mnLeft  = l
+    , mnRight = r
+    , mnRoot  = mkBranchRootHash (mnRoot l) (mnRoot r)
     }
 
-mkBranchRootHash :: MerkleSignature a -- ^ left merkle root
-                 -> MerkleSignature a -- ^ right merkle root
-                 -> MerkleSignature a
-mkBranchRootHash (MerkleSignature (AbstractHash hl) sl)
-                 (MerkleSignature (AbstractHash hr) sr)
-   = MerkleSignature
-   (hash $ toLazyByteString $ mconcat
-      [ word32LE sl
-      , word32LE sr
-      , byteString (convert hl)
-      , byteString (convert hr) ])
-   (sl + sr)
+-- | Concatenates two @'MerkleSignature'@s and calculates another
+-- @'MerkleSignature'@ on top of them.
+mkBranchRootHash
+    :: MerkleSignature a -- ^ left merkle root
+    -> MerkleSignature a -- ^ right merkle root
+    -> MerkleSignature a
+mkBranchRootHash (MerkleSignature sl hl) (MerkleSignature sr hr) =
+    MerkleSignature
+    (sl + sr)
+    (hash $ toLazyByteString $ mconcat
+        [ word32LE sl
+        , byteString (convert hl)
+        , word32LE sr
+        , byteString (convert hr) ])
   where
     toLazyByteString :: Builder -> LBS.ByteString
-    toLazyByteString = Builder.toLazyByteStringWith (Builder.safeStrategy 1024 4096) mempty
-
--- | Smart constructor for MerkleTree.
+    toLazyByteString = Builder.toLazyByteStringWith
+        -- This bytestring is build only to be hashed, and then it's
+        -- discarded. @'untrimmedStrategy'@ is a recommended strategy
+        -- for this case.
+        -- All the stuff to hash fits into 72 bytes, but let's allocate
+        -- 128 in the first buffer just in case (we don't know inner @'Builder'@)
+        -- machinery good enough.
+        (Builder.untrimmedStrategy 128 Builder.smallChunkSize)
+        mempty
+
+-- | Construct a @'MerkleTree'@ over a @'Foldable'@ container.
 fromFoldable :: (HasHash a, Foldable t) => t a -> MerkleTree a
 fromFoldable = fromList . F.toList
 
--- | Smart constructor for MerkleTree.
+-- | Construct a @'MerkleTree'@ over a @'Container'@
 fromContainer :: (HasHash (Element t), Container t) => t -> MerkleTree (Element t)
 fromContainer = fromList . toList
 
+-- | Construct a @'MerkleTree'@ over a list.
 fromList :: HasHash a => [a] -> MerkleTree a
-fromList [] = MerkleEmpty
-fromList ls = MerkleTree (nodeFromList ls)
+fromList []       = MerkleEmpty
+fromList (a : as) = MerkleTree $ nodeFromList (a :| as)
```

Comment:
I'd appreciate if you make it part of `IsList`, I got so many :fire_engine: from this `fromList` at top level.

Reply:
Ah right, I missed the fact that any reasonable use of `foldMap` is with `Endo` as its monoid.